components/style/values/specified/grid.rs

```diff
@@ -301,3 +340,377 @@ impl<L: ToComputedValue> ToComputedValue for TrackSize<L> {
+                              -> Result<(TrackRepeat<LengthOrPercentage>, RepeatType), ()> {
+        input.try(|i| i.expect_function_matching("repeat")).and_then(|_| {
+            input.parse_nested_block(|input| {
+                let count = RepeatCount::parse(context, input)?;
+                input.expect_comma()?;
+
+                let is_auto = count == RepeatCount::AutoFit || count == RepeatCount::AutoFill;
+                let mut repeat_type = if is_auto {
+                    RepeatType::Auto
+                } else {    // <fixed-size> is a subset of <track_size>, so it should work for both
+                    RepeatType::Fixed
+                };
+
+                let mut current_names;
+                let mut variants = vec![];
+
+                loop {
+                    current_names = input.try(parse_line_names).unwrap_or(vec![]);
+                    if let Ok(track_size) = input.try(|i| TrackSize::parse(context, i)) {
+                        if !track_size.is_fixed() {
+                            if is_auto {
+                                return Err(())      // should be <fixed-size> for <auto-repeat>
+                            }
+
+                            if repeat_type == RepeatType::Fixed {
+                                repeat_type = RepeatType::Normal       // <track-size> for sure
+                            }
+                        }
+
+                        variants.push((current_names, track_size));
+                    } else {
+                        if variants.is_empty() {
+                            return Err(())      // expecting at least one <track-size>
+                        }
+
+                        break       // no more <track-size>, breaking
+                    }
+                }
+
+                let repeat = TrackRepeat {
+                    count: count,
+                    repeat_variants: variants,
+                    line_names_last: current_names,    // last set of <line-names>
+                };
+
+                Ok((repeat, repeat_type))
+            })
+        })
+    }
+}
+
+impl<L: ToCss> ToCss for TrackRepeat<L> {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        dest.write_str("repeat(")?;
+        self.count.to_css(dest)?;
+        dest.write_str(", ")?;
+
+        let total = self.repeat_variants.len();
+        for (i, &(ref names, ref size)) in self.repeat_variants.iter().enumerate() {
+            if !names.is_empty() {
+                dest.write_str("[")?;
+                dest.write_str(&names.join(" "))?;
+                dest.write_str("] ")?;
+            }
+
+            size.to_css(dest)?;
+            if i < total - 1 {
+                dest.write_str(" ")?;
+            }
+        }
+
+        if !self.line_names_last.is_empty() {
+            dest.write_str(" [")?;
+            dest.write_str(&self.line_names_last.join(" "))?;
```

Comment:
let's just iterate over line_names_last in order to avoid allocating? Seems easy enough.